src/Processors/QueryPlan/Optimizations/optimizeTree.cpp

```diff
@@ -54,8 +57,13 @@ void optimizeTree(QueryPlan::Node & root, QueryPlan::Nodes & nodes)
             if (!optimization.apply)
                 continue;
 
+            if (max_optimizations_to_apply && max_optimizations_to_apply < total_applied_optimizations)
```

Comment:
What is it?)

Reply:
I want to have some protection from having infinite loop of optimizations.
Probably I should set `total_applied_optimizations` to something big like 10000 and throw exception if it was achieved.